### PR TITLE
Improve qscript merging of Reduce.

### LINKED ISSUE #2186: Query returning zero results as keys are empty
There are two similar queries, one that work (A) and one that does not work (B)

*A*: Working query
```SELECT avg(COUNT(*)) as cnt, state as category, gender FROM `/loc/patients-mini.data` WHERE codes[*].desc like "%flu%" GROUP BY state, gender ORDER BY COUNT(*) DESC```

*B*: Query that does not work -  gives empty result set
```select avg(cnt) as measure,  state as catergory from (SELECT COUNT(*) as cnt, state, gender FROM `/loc/patients-mini.data` WHERE codes[*].desc like "%flu%" GROUP BY state, gender ORDER BY COUNT(*) DESC) as meh``` 

As it turns out there is a JOIN where left and right side of join are exact same sets of elements, but the functions that generate left and right key for the join give ```Constant(Arr(List()))```

Logical Plan for query *A*

```
Let('__tmp0)
├─ Let('__tmp1)
│  ├─ Read(/loc/patients-mini.data)
│  ╰─ Typecheck(Obj(Map[], Some(Top)))
│     ├─ Free('__tmp1)
│     ├─ Free('__tmp1)
│     ╰─ Constant(NA)
╰─ Let('__tmp2)
   ├─ Invoke(Filter)
   │  ├─ Free('__tmp0)
   │  ╰─ Let('__tmp3)
   │     ├─ Invoke(ObjectProject)
   │     │  ├─ Invoke(FlattenArray)
   │     │  │  ╰─ Let('__tmp4)
   │     │  │     ├─ ObjectProject(Free('__tmp0){"codes"})
   │     │  │     ╰─ Typecheck(FlexArr(0, None, Obj(Map[], Some(Top))))
   │     │  │        ├─ Free('__tmp4)
   │     │  │        ├─ Free('__tmp4)
   │     │  │        ╰─ Constant(Arr(List(NA)))
   │     │  ╰─ Str("desc")
   │     ╰─ Typecheck(Str)
   │        ├─ Free('__tmp3)
   │        ├─ Invoke(Search)
   │        │  ├─ Free('__tmp3)
   │        │  ├─ Str("^.*flu.*$")
   │        │  ╰─ Constant(Bool(false))
   │        ╰─ Constant(NA)
   ╰─ Let('__tmp5)
      ├─ Invoke(GroupBy)
      │  ├─ Free('__tmp2)
      │  ╰─ Invoke(ArrayConcat)
      │     ├─ Invoke(MakeArray)
      │     │  ╰─ ObjectProject(Free('__tmp2){"state"})
      │     ╰─ Invoke(MakeArray)
      │        ╰─ ObjectProject(Free('__tmp2){"gender"})
      ╰─ Let('__tmp6)
         ├─ Invoke(Squash)
         │  ╰─ Invoke(ObjectConcat)
         │     ├─ Invoke(ObjectConcat)
         │     │  ├─ Invoke(ObjectConcat)
         │     │  │  ├─ Invoke(MakeObject)
         │     │  │  │  ├─ Str("cnt")
         │     │  │  │  ╰─ Invoke(Avg)
         │     │  │  │     ╰─ Invoke(Count)
         │     │  │  │        ╰─ Free('__tmp5)
         │     │  │  ╰─ Invoke(MakeObject)
         │     │  │     ├─ Str("category")
         │     │  │     ╰─ Invoke(Arbitrary)
         │     │  │        ╰─ ObjectProject(Free('__tmp5){"state"})
         │     │  ╰─ Invoke(MakeObject)
         │     │     ├─ Str("gender")
         │     │     ╰─ Invoke(Arbitrary)
         │     │        ╰─ ObjectProject(Free('__tmp5){"gender"})
         │     ╰─ Invoke(MakeObject)
         │        ├─ Str("__sd__0")
         │        ╰─ Invoke(Count)
         │           ╰─ Free('__tmp5)
         ╰─ Invoke(Take)
            ├─ Invoke(DeleteField)
            │  ├─ Sort
            │  │  ├─ Free('__tmp6)
            │  │  ├─ ObjectProject(Free('__tmp6){"__sd__0"})
            │  │  ╰─ SortDir(Descending)
            │  ╰─ Str("__sd__0")
            ╰─ Constant(Int(11))
```

Logical Plan for query *B*

```
Let('__tmp0)
├─ Let('__tmp1)
│  ├─ Read(/loc/patients-mini.data)
│  ╰─ Typecheck(Obj(Map[], Some(Top)))
│     ├─ Free('__tmp1)
│     ├─ Free('__tmp1)
│     ╰─ Constant(NA)
╰─ Let('__tmp2)
   ├─ Invoke(Filter)
   │  ├─ Free('__tmp0)
   │  ╰─ Let('__tmp3)
   │     ├─ Invoke(ObjectProject)
   │     │  ├─ Invoke(FlattenArray)
   │     │  │  ╰─ Let('__tmp4)
   │     │  │     ├─ ObjectProject(Free('__tmp0){"codes"})
   │     │  │     ╰─ Typecheck(FlexArr(0, None, Obj(Map[], Some(Top))))
   │     │  │        ├─ Free('__tmp4)
   │     │  │        ├─ Free('__tmp4)
   │     │  │        ╰─ Constant(Arr(List(NA)))
   │     │  ╰─ Str("desc")
   │     ╰─ Typecheck(Str)
   │        ├─ Free('__tmp3)
   │        ├─ Invoke(Search)
   │        │  ├─ Free('__tmp3)
   │        │  ├─ Str("^.*flu.*$")
   │        │  ╰─ Constant(Bool(false))
   │        ╰─ Constant(NA)
   ╰─ Let('__tmp5)
      ├─ Invoke(GroupBy)
      │  ├─ Free('__tmp2)
      │  ╰─ Invoke(ArrayConcat)
      │     ├─ Invoke(MakeArray)
      │     │  ╰─ ObjectProject(Free('__tmp2){"state"})
      │     ╰─ Invoke(MakeArray)
      │        ╰─ ObjectProject(Free('__tmp2){"gender"})
      ╰─ Let('__tmp6)
         ├─ Invoke(Squash)
         │  ╰─ Invoke(ObjectConcat)
         │     ├─ Invoke(ObjectConcat)
         │     │  ├─ Invoke(MakeObject)
         │     │  │  ├─ Str("cnt")
         │     │  │  ╰─ Invoke(Count)
         │     │  │     ╰─ Free('__tmp5)
         │     │  ╰─ Invoke(MakeObject)
         │     │     ├─ Str("state")
         │     │     ╰─ Invoke(Arbitrary)
         │     │        ╰─ ObjectProject(Free('__tmp5){"state"})
         │     ╰─ Invoke(MakeObject)
         │        ├─ Str("gender")
         │        ╰─ Invoke(Arbitrary)
         │           ╰─ ObjectProject(Free('__tmp5){"gender"})
         ╰─ Let('__tmp7)
            ├─ Sort
            │  ├─ Free('__tmp6)
            │  ├─ ObjectProject(Free('__tmp6){"cnt"})
            │  ╰─ SortDir(Descending)
            ╰─ Invoke(Take)
               ├─ Invoke(Squash)
               │  ╰─ Invoke(ObjectConcat)
               │     ├─ Invoke(MakeObject)
               │     │  ├─ Str("measure")
               │     │  ╰─ Invoke(Avg)
               │     │     ╰─ Let('__tmp8)
               │     │        ├─ ObjectProject(Free('__tmp7){"cnt"})
               │     │        ╰─ Typecheck(Int, Dec, or Interval)
               │     │           ├─ Free('__tmp8)
               │     │           ├─ Free('__tmp8)
               │     │           ╰─ Constant(NA)
               │     ╰─ Invoke(MakeObject)
               │        ├─ Str("catergory")
               │        ╰─ ObjectProject(Free('__tmp7){"state"})
               ╰─ Constant(Int(11))
```

### PR DESCRIPTION
**Pull request by alissapajer**
Fixes #2186

### CODE REVIEW

**Comment by alissapajer** [comment]
@drostron hold of on merging this until I give the go ahead. We might need to PR this against backport instead of master.

**Comment by alissapajer** [comment]
@drostron ok, talked with @zeeiyerWork and this slated for 4.2.2 so master is the correct target for merge.

File: it/src/main/resources/tests/demo/nestedSelectWithReduce.test [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        "mongodb_3_2": "pending",
+        "mongodb_q_3_2": "pending",
+        "postgresql": "pending"
+    },
+    "data": "patients.data",
+    "query": "SELECT AVG(cnt) as measure, state as category FROM
+                (SELECT COUNT(*) as cnt, state, gender FROM patients
+                WHERE codes[*].desc LIKE \"%flu%\"
+                GROUP BY state, gender
+                ORDER BY COUNT(*) DESC) as meh",
```
Comment (alissapajer) [inline it/src/main/resources/tests/demo/nestedSelectWithReduce.test]: @sellout interestingly, this query does not complete on old mongo - it gives a "failed to merge" error.

Comment (sellout) [inline it/src/main/resources/tests/demo/nestedSelectWithReduce.test]: I’m not surprised – old Mongo breaks down pretty quickly on anything that does much with LeftShift ior autojoin.

**Pull request merged**